src/ImageSharp/IO/ChunkedMemoryStream.cs

```diff
@@ -507,48 +402,82 @@ private void EnsureNotDisposed()
     }
 
     [MethodImpl(MethodImplOptions.NoInlining)]
-    private static void ThrowDisposed() => throw new ObjectDisposedException(null, "The stream is closed.");
+    private static void ThrowDisposed() => throw new ObjectDisposedException(nameof(ChunkedMemoryStream), "The stream is closed.");
 
-    [MethodImpl(MethodImplOptions.NoInlining)]
-    private static void ThrowArgumentOutOfRange(string value) => throw new ArgumentOutOfRangeException(value);
+    private sealed class MemoryChunkBuffer : IEnumerable<MemoryChunk>, IDisposable
+    {
+        private readonly List<MemoryChunk> memoryChunks = new();
+        private readonly MemoryAllocator allocator;
+        private readonly int allocatorCapacity;
+        private bool isDisposed;
 
-    [MethodImpl(MethodImplOptions.NoInlining)]
-    private static void ThrowInvalidSeek() => throw new ArgumentException("Invalid seek origin.");
+        public MemoryChunkBuffer(MemoryAllocator allocator)
+        {
+            this.allocatorCapacity = allocator.GetBufferCapacityInBytes();
+            this.allocator = allocator;
+        }
 
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private MemoryChunk AllocateMemoryChunk()
-    {
-        // Tweak our buffer sizes to take the minimum of the provided buffer sizes
-        // or the allocator buffer capacity which provides us with the largest
-        // available contiguous buffer size.
-        IMemoryOwner<byte> buffer = this.allocator.Allocate<byte>(Math.Min(this.allocatorCapacity, GetChunkSize(this.chunkCount++)));
+        public int ChunkCount => this.memoryChunks.Count;
+
+        public long Length { get; private set; }
 
-        return new MemoryChunk(buffer)
+        public MemoryChunk this[int index] => this.memoryChunks[index];
+
+        public void Expand()
         {
-            Next = null,
-            Length = buffer.Length()
-        };
-    }
+            IMemoryOwner<byte> buffer =
+                this.allocator.Allocate<byte>(Math.Min(this.allocatorCapacity, GetChunkSize(this.ChunkCount)));
 
-    private static void ReleaseMemoryChunks(MemoryChunk? chunk)
-    {
-        while (chunk != null)
+            MemoryChunk chunk = new(buffer)
+            {
+                Length = buffer.Length()
+            };
+
+            this.memoryChunks.Add(chunk);
+            this.Length += chunk.Length;
+        }
+
+        public void Dispose()
         {
-            chunk.Dispose();
-            chunk = chunk.Next;
+            this.Dispose(true);
+            GC.SuppressFinalize(this);
```

Comment:
I think it's not necessary to call `GC.SuppressFinalize(this)` in a sealed class with no finalizer. This would be to cover sub-types that have a finalizer (in scenarios where there is no finalizer defined directly on the type).

Reply:
Force of habit. Well spotted!